src/coreclr/jit/morph.cpp

```diff
@@ -13114,9 +13114,8 @@ GenTree* Compiler::fgOptimizeAddition(GenTreeOp* add)
         DEBUG_DESTROY_NODE(constTwo);
     }
 
-    // Fold (x + 0) - given it won't change the tree type to TYP_REF.
-    // TODO-Bug: this code will lose the GC-ness of a tree like "native int + byref(0)".
-    if (op2->IsIntegralConst(0) && ((add->TypeGet() == op1->TypeGet()) || !op1->TypeIs(TYP_REF)))
+    // Fold (x + 0) - given it won't change the tree type.
+    if (op2->IsIntegralConst(0) && (genActualType(add) == genActualType(op1)))
```

Comment:
Could we have something like TYP_LONG(N) + TYP_INT(0)?

Reply:
Technically that is invalid IR, practically we may see it (due to bugs / invalid IL in CLR tests / both).

I suppose this code should be correct for that case (well, not if it is `long + int => int`, but then that's actually invalid and I have never seen anything produce such forms).

Reply:
I don't think we will see such forms either, as you said, it is technically invalid.